### PR TITLE
Opti-CAM Attribution Method

### PR DESCRIPTION
**Pull request by Agustin-Picard**
## Opti-CAM Attribution Method

This PR contains the code implementing the [Opti-CAM attribution method](https://www.sciencedirect.com/science/article/pii/S1077314224001826?via%3Dihub)

The class that implements it follows the `WhiteBoxExplainer` interface, produces a CAM explanation, and optimizes the weights for each filter.

Some relevant tests are introduced to make sure that the losses, normalization, parameters, and shapes make sense.

### CODE REVIEW

File: xplique/attributions/opti_cam.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        weights
+            Weights of the feature maps (aka the CAM explanation) for the images.
+
+        Returns
+        -------
+        overlayed_images
+            Images with the explanation overlayed.
+        """
+        alpha = tf.nn.softmax(weights, axis=-1)
+        weighted_explanation = tf.reshape(
```
Comment (AntoninPoche) [inline xplique/attributions/opti_cam.py:175]: There should be a clearer way to write this. Maybe with `tf.einsum` or `tf.matvec`.

File: xplique/attributions/opti_cam.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        for x_batch, y_batch in tf.data.Dataset.from_tensor_slices((inputs, targets)).batch(
+                batch_size):
+            # initialize weights and optimization elements
+            current_batch_size = x_batch.shape[0]
+            weights = tf.Variable(
+                0.5 * tf.ones((current_batch_size, 1, 1, self.conv_layer.output.shape[-1])
+            ), trainable=True, dtype=tf.float32)
+            optimizer = tf.keras.optimizers.Adam(0.05)
+            for _ in range(self.n_iters):
+                with tf.GradientTape(watch_accessed_variables=False) as tape:
```
Comment (AntoninPoche) [inline xplique/attributions/opti_cam.py:130]: The explain function lacks comments.

File: tests/attributions/test_opti_cam.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    nb_labels = 10
+
+    for input_shape in input_shapes:
+        samples, labels = generate_data(input_shape, nb_labels, 100)
+        model = _generate_model(input_shape, nb_labels)
+
+        method = OptiCAM(model, -2)
+        outputs = method.explain(samples, labels)
+
+        assert samples.shape[:3] == outputs.shape[:3]
```
Comment (AntoninPoche) [inline tests/attributions/test_opti_cam.py:55]: You should also check that the outputs shape finish by a 1. There is a test for all methods where you should add OptiCAM. (tests/attributions/test_commons.py/test_data_types_shapes()).

**Review by AntoninPoche — CHANGES_REQUESTED** [review]
Nice job Agustin! The code looks good to me. But there is no documentation, and I left some comments.

Did you make an qualitative test?
